fetchFromGitProvider: init with `<pkg>.overrideAttrs` that dynamically switch fetchers

ShamrockLee opened the pull request:
This PR introduces `fetchFromGitProvider`, a general foundation for all other `fetchFrom<Git hosting provider>` constructors.

Not only can it choose between `fetchgit` (for Git-related features) and `fetchgit` (for hash stability) automatically, but it can optionally switch between `fetchgit` and `fetchzip` dynamically after `<pkg>.overrideAttrs` overriding based on the overridden arguments.

Such a black magic depends on the fact that `fetchurl`, `fetchzip`, and `fetchgit` *pass arguments unconditionally* (i.e., without the `lib.optionalAttrs` part of `stdenv.mkDerivation ({ ... } // lib.optionalAttrs myCondition { ... })`).
By making `fetchgithub` pass unconditionally to `fetchgit`, `fetchzip`, and `stdenvNoCC.mkDerivation`, dynamic switching is achievable.

The unconditional passing to `stdenvNoCC.mkDerivation` uses `null` as the generic hole-filling default value, and the unconditional passing to `fetchgit` is done in a depending PR listed below.

As the arguments passing to the bottom `stdenvNoCC.mkDerivation` is predictable, this PR utilise the `<build helper>.expectDrvArgs` attribute provided by `fetchurl`, `fetchzip`, and `fetchgit`.

Due to dynamic switching of fetchers being and the `null`-hole-filling argument set being larger than a simple argument set passed to individual fetchers, `fetchFromGitProvider` implements two ways of backend fetcher choosing -- `fetchFromGitProviderDynamic` and `fetchFromGitProviderSimple`, and construct the result derivation with `fetchFromGitProviderSimple` without the dynamic switching capability, and then attach the `<pkg>.overrideAttrs` from an alternative result derivation constructed with `fetchFromGitProviderDynamic`.

This chimera style of implementation might results in the `pkg != pkg.overrideAttrs { }` surprises, but it eliminates possible computational overhead (from +15% to <+%1).

Depending PRs:
- #553249
- #553251
- #553631

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform:
  - [x] x86_64-linux
  - [x] aarch64-linux
  - [ ] x86_64-darwin
  - [x] aarch64-darwin
- Tested, as applicable:
  - [ ] [NixOS tests] in [nixos/tests].
  - [ ] [Package tests] at `passthru.tests`.
  - [x] Tests in [lib/tests] or [pkgs/test] for functions and "core" functionality.
- [x] Ran `nixpkgs-review` on this PR. See [nixpkgs-review usage].
- [ ] Tested basic functionality of all binary files, usually in `./result/bin/`.
- Nixpkgs Release Notes
  - [ ] Package update: when the change is major or breaking.
- NixOS Release Notes
  - [ ] Module addition: when adding a new NixOS module.
  - [ ] Module update: when the change is significant.
- [x] Fits [CONTRIBUTING.md], [pkgs/README.md], [maintainers/README.md] and other READMEs.

[NixOS tests]: https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests
[Package tests]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests
[nixpkgs-review usage]: https://github.com/Mic92/nixpkgs-review#usage

[CONTRIBUTING.md]: https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md
[lib/tests]: https://github.com/NixOS/nixpkgs/blob/master/lib/tests
[maintainers/README.md]: https://github.com/NixOS/nixpkgs/blob/master/maintainers/README.md
[nixos/tests]: https://github.com/NixOS/nixpkgs/blob/master/nixos/tests
[pkgs/README.md]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md
[pkgs/test]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/test

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

ShamrockLee committed:
fetchFromGitProvider: init with a <pkg.overrideAttrs> that switches fetchers

ShamrockLee committed:
fetchFromGitHub: base on fetchFromGitProvider and lib.extendMkDerivation

ShamrockLee commented:
Now it passes all the existing tests, but the "CPU time increased by 14.5%" performance overhead is not quite acceptable.

I'll try to optimize it.

## Performance comparison

This compares the performance of this branch against its pull request base branch (e.g., 'master')

For further help please refer to: [ci/README.md](https://github.com/NixOS/nixpkgs/blob/master/ci/README.md)

### Unchanged values

| metric                          |   value |
|---------------------------------|---------|
| sizes.Attr[^sizes.Attr]         |      16 |
| sizes.Bindings[^sizes.Bindings] |      24 |
| sizes.Env[^sizes.Env]           |       8 |
| sizes.Value[^sizes.Value]       |      16 |

### Updated values

| metric                                          |     mean_before |      mean_after |      mean_diff |   mean_pct_change | p_value   |   t_stat |
|-------------------------------------------------|-----------------|-----------------|----------------|-------------------|-----------|----------|
| time.cpu[^time.cpu]                             |         26.2992 |         30.1918 |         3.8926 |           14.5303 | -         |  13.3068 |
| time.gc[^time.gc]                               |          2.0607 |          2.7813 |         0.7206 |           38.9145 | -         |  11.3923 |
| time.gcFraction[^time.gcFraction]               |          0.0755 |          0.0897 |         0.0142 |           20.7026 | -         |   9.1617 |
| gc.cycles[^gc.cycles]                           |          8.5769 |         10.3077 |         1.7308 |           20.9264 | -         |  14.4152 |
| gc.heapSize[^gc.heapSize]                       | 1846401181.5385 | 2085153870.7692 | 238752689.2308 |           14.4830 | -         |  10.0539 |
| gc.totalBytes[^gc.totalBytes]                   | 3481972329.2308 | 4381634855.0769 | 899662525.8462 |           26.6369 | -         |  13.9828 |
| envs.bytes[^envs.bytes]                         |  611916669.0769 |  701896202.9231 |  89979533.8462 |           15.2425 | -         |  14.2532 |
| list.bytes[^list.bytes]                         |   86394604.1538 |   90307111.2308 |   3912507.0769 |            4.6224 | -         |  13.9794 |
| sets.bytes[^sets.bytes]                         | 1327595539.8462 | 1952547122.7692 | 624951582.9231 |           48.9251 | -         |  13.9015 |
| symbols.bytes[^symbols.bytes]                   |    1452974.2308 |    1453715.2308 |       741.0000 |            0.0545 | -         | 549.8722 |
| values.bytes[^values.bytes]                     |  825376726.4615 |  950151677.2308 | 124774950.7692 |           15.8872 | -         |  14.1941 |
| envs.number[^envs.number]                       |   30747166.1923 |   35280877.4231 |   4533711.2308 |           15.2844 | -         |  14.2188 |
| nrAvoided[^nrAvoided]                           |   37278251.5385 |   40999689.9423 |   3721438.4038 |           10.1547 | -         |  14.0338 |
| nrExprs[^nrExprs]                               |    1564427.7692 |    1564889.3077 |       461.5385 |            0.0313 | -         | 618.4658 |
| nrFunctionCalls[^nrFunctionCalls]               |   27687666.7308 |   31735632.5000 |   4047965.7692 |           15.1248 | -         |  14.1329 |
| nrLookups[^nrLookups]                           |   13612368.8462 |   15983994.3462 |   2371625.5000 |           18.4031 | -         |  13.9755 |
| nrOpUpdateValuesCopied[^nrOpUpdateValuesCopied] |   42290049.1923 |   62196596.3846 |  19906547.1923 |           49.3418 | -         |  13.9000 |
| nrOpUpdates[^nrOpUpdates]                       |    4864289.5577 |    6478620.5192 |   1614330.9615 |           33.5087 | -         |  14.3086 |
| nrPrimOpCalls[^nrPrimOpCalls]                   |   16073393.9231 |   17056812.2115 |    983418.2885 |            6.1660 | -         |  14.1666 |
| nrThunks[^nrThunks]                             |   39150908.3077 |   43091064.4038 |   3940156.0962 |           10.4801 | -         |  14.5125 |
| sets.number[^sets.number]                       |    6759438.0577 |    8456160.2692 |   1696722.2115 |           26.1742 | -         |  14.0699 |
| symbols.number[^symbols.number]                 |     117413.7308 |     117452.3462 |        38.6154 |            0.0342 | -         | 373.7618 |
| values.number[^values.number]                   |   51586045.4038 |   59384479.8269 |   7798434.4231 |           15.8872 | -         |  14.1941 |
| envs.elements[^envs.elements]                   |   45742417.4423 |   52456147.9423 |   6713730.5000 |           15.2184 | -         |  14.2764 |
| list.concats[^list.concats]                     |    2202940.2885 |    2245403.3077 |     42463.0192 |            2.0064 | -         |  13.8849 |
| list.elements[^list.elements]                   |   10799325.5192 |   11288388.9038 |    489063.3846 |            4.6224 | -         |  13.9794 |
| sets.elements[^sets.elements]                   |   72835564.1538 |  109349954.7692 |  36514390.6154 |           52.2253 | -         |  13.8898 |

[^time.cpu]: Number of seconds of CPU time accounted by the OS to the Nix evaluator process. On UNIX systems, this comes from [`getrusage(RUSAGE_SELF)`](https://man7.org/linux/man-pages/man2/getrusage.2.html).
[^time.gc]: Number of seconds of CPU time accounted by the Boehm garbage collector to performing GC.
[^time.gcFraction]: What fraction of the total CPU time is accounted towards performing GC.
[^gc.cycles]: Number of times garbage collection has been performed.
[^gc.heapSize]: Size in bytes of the garbage collector heap.
[^gc.totalBytes]: Size in bytes of all allocations in the garbage collector.
[^envs.bytes]: Size in bytes of all `Env` objects allocated by the Nix evaluator. These are almost exclusively created by [`nix-env`](https://nix.dev/manual/nix/stable/command-ref/nix-env.html).
[^list.bytes]: Size in bytes of all [lists](https://nix.dev/manual/nix/stable/language/syntax.html#list-literal) allocated by the Nix evaluator.
[^sets.bytes]: Size in bytes of all [attrsets](https://nix.dev/manual/nix/stable/language/syntax.html#list-literal) allocated by the Nix evaluator.
[^symbols.bytes]: Size in bytes of all items in the Nix evaluator symbol table.
[^values.bytes]: Size in bytes of all values allocated by the Nix evaluator.
[^envs.number]: The count of all `Env` objects allocated.
[^nrAvoided]: The number of thunks avoided being created.
[^nrExprs]: The number of expression objects ever created.
[^nrFunctionCalls]: The number of function calls ever made.
[^nrLookups]: The number of lookups into an attrset ever made.
[^nrOpUpdateValuesCopied]: The number of attrset values copied in the process of merging attrsets.
[^nrOpUpdates]: The number of attrsets merge operations (`//`) performed.
[^nrPrimOpCalls]: The number of function calls to primops (Nix builtins) ever made.
[^nrThunks]: The number of [thunks](https://nix.dev/manual/nix/latest/language/evaluation.html#laziness) ever made. A thunk is a delayed computation, represented by an expression reference and a closure.
[^sets.number]: The number of attrsets ever made.
[^symbols.number]: The number of symbols ever added to the symbol table.
[^values.number]: The number of values ever made.
[^envs.elements]: The number of values contained within an `Env` object.
[^list.concats]: The number of list concatenation operations (`++`) performed.
[^list.elements]: The number of values contained within a list.
[^sets.elements]: The number of values contained within an attrset.
[^sizes.Attr]: Size in bytes of the `Attr` type.
[^sizes.Bindings]: Size in bytes of the `Bindings` type.
[^sizes.Env]: Size in bytes of the `Env` type.
[^sizes.Value]: Size in bytes of the `Value` type.

ShamrockLee commented:
After refactoring `accumulateConstructorMetadata`, the performance overhead (CPU time) with aggressive dynamic switching enabled drop from 9%-10% to 6%-10%.

Unfortunately, the performance overhead *without* aggressive dynamic switching (just restructuring with `lib.extendMkDerivation` while preserving the one-time switching functionality) still has 4~6% performance overhead.

I unrolled several call-time determined `{ fetchgit = ...; fetchzip = ...; }` pairs into two variables for performance gains. This contributes to about -1% CPU times in full evaluation.

xokdvium commented:
The 5% increase in memory usage seems pretty steep. Any idea where it is coming from? CPU time is not that big of a concern all things considered, but hiking memory usage is quite prohibitive.

ShamrockLee commented:
> Any idea where it is coming from?

I just checked all the depending PRs and confirmed that the memory overhead comes from changes provided by this PR.

It uses quite a few intermediate attribute sets at call time. I'll see how much I could optimize out.

ShamrockLee committed:
lib.fetchers.normalizeHash: move assertions down to outputHash and outputHashAlgo values

ShamrockLee committed:
fetchgit: expose resolveNullableFetchGit for fetchFromGitProvider

ShamrockLee committed:
lib.extendMkDerivation: optionally calculate expectDrvArgs

ShamrockLee committed:
fetchurl: provide expectDrvArgs

ShamrockLee committed:
fetchzip: provide expectDrvArgs

ShamrockLee committed:
fetchgit: provide expectDrvArgs

ShamrockLee committed:
prefer-remote-fetch: preserve callable attributes

ShamrockLee committed:
prefer-remote-fetch: recalculate expectDrvArgs if the original fetcher provides one

ShamrockLee commented:
While `gc.heapSize` still has a 1.7% increase, the CPU time change is now negligible.

Performance issues are no longer blocking this implementation.

## Unchanged values

| metric                          |   value |
|---------------------------------|---------|
| sizes.Attr[^sizes.Attr]         |      16 |
| sizes.Bindings[^sizes.Bindings] |      24 |
| sizes.Env[^sizes.Env]           |       8 |
| sizes.Value[^sizes.Value]       |      16 |

## Updated values

| metric                                          |     mean_before |      mean_after |     mean_diff |   mean_pct_change |   p_value |    t_stat |
|-------------------------------------------------|-----------------|-----------------|---------------|-------------------|-----------|-----------|
| time.cpu[^time.cpu]                             |         25.2424 |         24.7707 |       -0.4717 |           -1.9317 |    0.0155 |   -2.5348 |
| time.gc[^time.gc]                               |          2.1757 |          2.5688 |        0.3932 |           12.0564 |    0.1812 |    1.3619 |
| time.gcFraction[^time.gcFraction]               |          0.0802 |          0.0921 |        0.0119 |           13.4776 |    0.1473 |    1.4792 |
| gc.cycles[^gc.cycles]                           |          7.3590 |          7.4615 |        0.1026 |            1.2048 |    0.0439 |    2.0840 |
| gc.heapSize[^gc.heapSize]                       | 2230219277.1282 | 2261520068.9231 | 31300791.7949 |            1.7397 |    0.0003 |    4.0133 |
| gc.totalBytes[^gc.totalBytes]                   | 3650405865.8462 | 3709375296.4103 | 58969430.5641 |            1.7637 |    -      |   12.7874 |
| envs.bytes[^envs.bytes]                         |  465557119.7949 |  477492616.8205 | 11935497.0256 |            2.8962 |    -      |   12.5975 |
| list.bytes[^list.bytes]                         |   77280173.3333 |   76804717.9487 |  -475455.3846 |           -0.7247 |    -      |   -9.6621 |
| sets.bytes[^sets.bytes]                         | 1621171556.3077 | 1657492123.0769 | 36320566.7692 |            2.4158 |    -      |   12.5845 |
| symbols.bytes[^symbols.bytes]                   |    1669699.6667 |    1670470.1282 |      770.4615 |            0.0509 |    -      |  555.8128 |
| values.bytes[^values.bytes]                     |  730957514.6667 |  739379564.3077 |  8422049.6410 |            1.3297 |    -      |   12.5672 |
| envs.number[^envs.number]                       |   21102755.5641 |   21791892.1026 |   689136.5385 |            3.7504 |    -      |   12.5925 |
| nrAvoided[^nrAvoided]                           |   28907286.6923 |   29449859.3590 |   542572.6667 |            2.0526 |    -      |   12.6177 |
| nrExprs[^nrExprs]                               |    1609494.0513 |    1610273.3590 |      779.3077 |            0.0518 |    -      | 4507.0617 |
| nrFunctionCalls[^nrFunctionCalls]               |   18218504.7949 |   18729027.8974 |   510523.1026 |            3.2221 |    -      |   12.5918 |
| nrLookups[^nrLookups]                           |    9863997.2564 |    9995299.5385 |   131302.2821 |            1.5149 |    -      |   12.4301 |
| nrOpUpdateValuesCopied[^nrOpUpdateValuesCopied] |   51243828.0769 |   52025583.9231 |   781755.8462 |            1.6559 |    -      |   12.5193 |
| nrOpUpdates[^nrOpUpdates]                       |    4462672.3077 |    4616460.0000 |   153787.6923 |            3.6521 |    -      |   12.5938 |
| nrPrimOpCalls[^nrPrimOpCalls]                   |    9663093.5128 |    9746543.5641 |    83450.0513 |            0.9680 |    -      |   12.6399 |
| nrThunks[^nrThunks]                             |   28581982.8974 |   29036095.8205 |   454112.9231 |            1.8000 |    -      |   12.5922 |
| sets.number[^sets.number]                       |    6788218.7949 |    6954411.6410 |   166192.8462 |            2.6700 |    -      |   12.6005 |
| symbols.number[^symbols.number]                 |     127865.6154 |     127909.0256 |       43.4103 |            0.0357 |    -      |  319.0486 |
| values.number[^values.number]                   |   45684844.6667 |   46211222.7692 |   526378.1026 |            1.3297 |    -      |   12.5672 |
| envs.elements[^envs.elements]                   |   37091884.4103 |   37894685.0000 |   802800.5897 |            2.4241 |    -      |   12.6007 |
| list.concats[^list.concats]                     |    1889515.6667 |    1880769.7179 |    -8745.9487 |           -0.5092 |    -      |  -12.4775 |
| list.elements[^list.elements]                   |    9660021.6667 |    9600589.7436 |   -59431.9231 |           -0.7247 |    -      |   -9.6621 |
| sets.elements[^sets.elements]                   |   91140894.0769 |   93161640.2308 |  2020746.1538 |            2.3954 |    -      |   12.5825 |

[^time.cpu]: Number of seconds of CPU time accounted by the OS to the Nix evaluator process. On UNIX systems, this comes from [`getrusage(RUSAGE_SELF)`](https://man7.org/linux/man-pages/man2/getrusage.2.html).
[^time.gc]: Number of seconds of CPU time accounted by the Boehm garbage collector to performing GC.
[^time.gcFraction]: What fraction of the total CPU time is accounted towards performing GC.
[^gc.cycles]: Number of times garbage collection has been performed.
[^gc.heapSize]: Size in bytes of the garbage collector heap.
[^gc.totalBytes]: Size in bytes of all allocations in the garbage collector.
[^envs.bytes]: Size in bytes of all `Env` objects allocated by the Nix evaluator. These are almost exclusively created by [`nix-env`](https://nix.dev/manual/nix/stable/command-ref/nix-env.html).
[^list.bytes]: Size in bytes of all [lists](https://nix.dev/manual/nix/stable/language/syntax.html#list-literal) allocated by the Nix evaluator.
[^sets.bytes]: Size in bytes of all [attrsets](https://nix.dev/manual/nix/stable/language/syntax.html#list-literal) allocated by the Nix evaluator.
[^symbols.bytes]: Size in bytes of all items in the Nix evaluator symbol table.
[^values.bytes]: Size in bytes of all values allocated by the Nix evaluator.
[^envs.number]: The count of all `Env` objects allocated.
[^nrAvoided]: The number of thunks avoided being created.
[^nrExprs]: The number of expression objects ever created.
[^nrFunctionCalls]: The number of function calls ever made.
[^nrLookups]: The number of lookups into an attrset ever made.
[^nrOpUpdateValuesCopied]: The number of attrset values copied in the process of merging attrsets.
[^nrOpUpdates]: The number of attrsets merge operations (`//`) performed.
[^nrPrimOpCalls]: The number of function calls to primops (Nix builtins) ever made.
[^nrThunks]: The number of [thunks](https://nix.dev/manual/nix/latest/language/evaluation.html#laziness) ever made. A thunk is a delayed computation, represented by an expression reference and a closure.
[^sets.number]: The number of attrsets ever made.
[^symbols.number]: The number of symbols ever added to the symbol table.
[^values.number]: The number of values ever made.
[^envs.elements]: The number of values contained within an `Env` object.
[^list.concats]: The number of list concatenation operations (`++`) performed.
[^list.elements]: The number of values contained within a list.
[^sets.elements]: The number of values contained within an attrset.
[^sizes.Attr]: Size in bytes of the `Attr` type.
[^sizes.Bindings]: Size in bytes of the `Bindings` type.
[^sizes.Env]: Size in bytes of the `Env` type.
[^sizes.Value]: Size in bytes of the `Value` type.

nixpkgs-review-gha commented:
## :x: `nixpkgs-review` result

Generated using [`nixpkgs-review-gha`](https://github.com/Defelo/nixpkgs-review-gha) ([`71d8654`](https://github.com/Defelo/nixpkgs-review-gha/commit/71d865464aa149e8ac87cc23305ece0506089a89))

Command: `nixpkgs-review pr 462034`
Commit: [`ec87e96db2cd03ef8572184660172e3e1ef2f7da`](https://github.com/NixOS/nixpkgs/commit/ec87e96db2cd03ef8572184660172e3e1ef2f7da) ([subsequent changes](https://github.com/NixOS/nixpkgs/compare/ec87e96db2cd03ef8572184660172e3e1ef2f7da..pull/462034/head))
Merge: [`04d529280b44a4028a92865b8a7af1c985377536`](https://github.com/NixOS/nixpkgs/commit/04d529280b44a4028a92865b8a7af1c985377536)

Triggered by @ShamrockLee (44064051)
Logs: https://github.com/ShamrockLee/nixpkgs-review-gha/actions/runs/31965361651/attempts/1



---
### `x86_64-linux`
<details><summary>:fast_forward: 2 packages blacklisted:</summary><ul>
<li>nixos-install-tools</li>
<li>tests.nixos-functions.nixos-test</li>
</ul></details><details><summary>:x: 1 package failed to build:</summary><ul>
<li>tests.testers.runCommand.bork</li>
</ul></details><details><summary>:white_check_mark: 55 packages built:</summary><ul>
<li>extra-container</li>
<li>nixos-container</li>
<li>nixpkgs-manual</li>
<li>tests.fetchFromBitbucket.withEncodedWhitespaceGit</li>
<li>tests.fetchFromGitHub.describe-tag</li>
<li>tests.fetchFromGitHub.dumb-http-signed-tag</li>
<li>tests.fetchFromGitHub.fetchTags</li>
<li>tests.fetchFromGitHub.leave-git</li>
<li>tests.fetchFromGitHub.rootDir</li>
<li>tests.fetchFromGitHub.simple</li>
<li>tests.fetchFromGitHub.simple-tag</li>
<li>tests.fetchFromGitHub.sparseCheckout</li>
<li>tests.fetchFromGitHub.sparseCheckoutNonConeMode</li>
<li>tests.fetchFromGitHub.submodule-deep</li>
<li>tests.fetchFromGitHub.submodule-leave-git</li>
<li>tests.fetchFromGitHub.submodule-leave-git-deep</li>
<li>tests.fetchFromGitHub.submodule-simple</li>
<li>tests.fetchFromHuggingFace.rootDir</li>
<li>tests.fetchFromHuggingFace.simple</li>
<li>tests.fetchgit.collect-rev</li>
<li>tests.fetchgit.describe-tag</li>
<li>tests.fetchgit.dumb-http-signed-tag</li>
<li>tests.fetchgit.fetchTags</li>
<li>tests.fetchgit.leave-git</li>
<li>tests.fetchgit.prefetch-git-no-add-path</li>
<li>tests.fetchgit.rootDir</li>
<li>tests.fetchgit.simple</li>
<li>tests.fetchgit.simple-tag</li>
<li>tests.fetchgit.sparseCheckout</li>
<li>tests.fetchgit.sparseCheckoutNonConeMode</li>
<li>tests.fetchgit.submodule-deep</li>
<li>tests.fetchgit.submodule-leave-git</li>
<li>tests.fetchgit.submodule-leave-git-deep</li>
<li>tests.fetchgit.submodule-revision-count</li>
<li>tests.fetchgit.submodule-simple</li>
<li>tests.fetchgit.withGitConfig</li>
<li>tests.fetchpatch2.decode</li>
<li>tests.fetchpatch2.fileWithApostrophe</li>
<li>tests.fetchpatch2.fileWithSpace</li>
<li>tests.fetchpatch2.full</li>
<li>tests.fetchpatch2.hunks</li>
<li>tests.fetchpatch2.relative</li>
<li>tests.fetchpatch2.simple</li>
<li>tests.fetchurl.flag-appending-netrcPhase-curlOpts</li>
<li>tests.fetchurl.flag-appending-netrcPhase-curlOptsList</li>
<li>tests.fetchurl.header</li>
<li>tests.haskell.cabalSdist.assumptionLocalHasDirectReference</li>
<li>tests.haskell.cabalSdist.localHasNoDirectReference</li>
<li>tests.lib-tests</li>
<li>tests.pnpm.pnpm-fixup-state-db</li>
<li>tests.pnpm.pnpm-workspaces.legacy</li>
<li>tests.pnpm.pnpm-workspaces.structured</li>
<li>tests.pnpm.pnpm_11_v4</li>
<li>tests.prefer-remote-fetch.fetchFromGitHub</li>
<li>tests.prefer-remote-fetch.fetchgit</li>
</ul></details>

---
### `aarch64-linux`
<details><summary>:fast_forward: 2 packages blacklisted:</summary><ul>
<li>nixos-install-tools</li>
<li>tests.nixos-functions.nixos-test</li>
</ul></details><details><summary>:x: 1 package failed to build:</summary><ul>
<li>tests.testers.runCommand.bork</li>
</ul></details><details><summary>:white_check_mark: 55 packages built:</summary><ul>
<li>extra-container</li>
<li>nixos-container</li>
<li>nixpkgs-manual</li>
<li>tests.fetchFromBitbucket.withEncodedWhitespaceGit</li>
<li>tests.fetchFromGitHub.describe-tag</li>
<li>tests.fetchFromGitHub.dumb-http-signed-tag</li>
<li>tests.fetchFromGitHub.fetchTags</li>
<li>tests.fetchFromGitHub.leave-git</li>
<li>tests.fetchFromGitHub.rootDir</li>
<li>tests.fetchFromGitHub.simple</li>
<li>tests.fetchFromGitHub.simple-tag</li>
<li>tests.fetchFromGitHub.sparseCheckout</li>
<li>tests.fetchFromGitHub.sparseCheckoutNonConeMode</li>
<li>tests.fetchFromGitHub.submodule-deep</li>
<li>tests.fetchFromGitHub.submodule-leave-git</li>
<li>tests.fetchFromGitHub.submodule-leave-git-deep</li>
<li>tests.fetchFromGitHub.submodule-simple</li>
<li>tests.fetchFromHuggingFace.rootDir</li>
<li>tests.fetchFromHuggingFace.simple</li>
<li>tests.fetchgit.collect-rev</li>
<li>tests.fetchgit.describe-tag</li>
<li>tests.fetchgit.dumb-http-signed-tag</li>
<li>tests.fetchgit.fetchTags</li>
<li>tests.fetchgit.leave-git</li>
<li>tests.fetchgit.prefetch-git-no-add-path</li>
<li>tests.fetchgit.rootDir</li>
<li>tests.fetchgit.simple</li>
<li>tests.fetchgit.simple-tag</li>
<li>tests.fetchgit.sparseCheckout</li>
<li>tests.fetchgit.sparseCheckoutNonConeMode</li>
<li>tests.fetchgit.submodule-deep</li>
<li>tests.fetchgit.submodule-leave-git</li>
<li>tests.fetchgit.submodule-leave-git-deep</li>
<li>tests.fetchgit.submodule-revision-count</li>
<li>tests.fetchgit.submodule-simple</li>
<li>tests.fetchgit.withGitConfig</li>
<li>tests.fetchpatch2.decode</li>
<li>tests.fetchpatch2.fileWithApostrophe</li>
<li>tests.fetchpatch2.fileWithSpace</li>
<li>tests.fetchpatch2.full</li>
<li>tests.fetchpatch2.hunks</li>
<li>tests.fetchpatch2.relative</li>
<li>tests.fetchpatch2.simple</li>
<li>tests.fetchurl.flag-appending-netrcPhase-curlOpts</li>
<li>tests.fetchurl.flag-appending-netrcPhase-curlOptsList</li>
<li>tests.fetchurl.header</li>
<li>tests.haskell.cabalSdist.assumptionLocalHasDirectReference</li>
<li>tests.haskell.cabalSdist.localHasNoDirectReference</li>
<li>tests.lib-tests</li>
<li>tests.pnpm.pnpm-fixup-state-db</li>
<li>tests.pnpm.pnpm-workspaces.legacy</li>
<li>tests.pnpm.pnpm-workspaces.structured</li>
<li>tests.pnpm.pnpm_11_v4</li>
<li>tests.prefer-remote-fetch.fetchFromGitHub</li>
<li>tests.prefer-remote-fetch.fetchgit</li>
</ul></details>

---
### `aarch64-darwin` (sandbox = relaxed)
<details><summary>:x: 1 package failed to build:</summary><ul>
<li>tests.testers.runCommand.bork</li>
</ul></details><details><summary>:white_check_mark: 79 packages built:</summary><ul>
<li>nixpkgs-manual</li>
<li>tests.fetchFirefoxAddon.simple (tests.fetchFirefoxAddon.overridden-source)</li>
<li>tests.fetchFromBitbucket.withEncodedWhitespace</li>
<li>tests.fetchFromBitbucket.withEncodedWhitespaceGit</li>
<li>tests.fetchFromBitbucket.withoutWhitespace</li>
<li>tests.fetchFromGitHub.describe-tag</li>
<li>tests.fetchFromGitHub.dumb-http-signed-tag</li>
<li>tests.fetchFromGitHub.fetchTags</li>
<li>tests.fetchFromGitHub.leave-git</li>
<li>tests.fetchFromGitHub.rootDir</li>
<li>tests.fetchFromGitHub.simple</li>
<li>tests.fetchFromGitHub.simple-tag</li>
<li>tests.fetchFromGitHub.sparseCheckout</li>
<li>tests.fetchFromGitHub.sparseCheckoutNonConeMode</li>
<li>tests.fetchFromGitHub.submodule-deep</li>
<li>tests.fetchFromGitHub.submodule-leave-git</li>
<li>tests.fetchFromGitHub.submodule-leave-git-deep</li>
<li>tests.fetchFromGitHub.submodule-simple</li>
<li>tests.fetchFromHuggingFace.rootDir</li>
<li>tests.fetchFromHuggingFace.simple</li>
<li>tests.fetchNextcloudApp.simple-sha256</li>
<li>tests.fetchNextcloudApp.simple-sha512</li>
<li>tests.fetchgit.collect-rev</li>
<li>tests.fetchgit.describe-tag</li>
<li>tests.fetchgit.dumb-http-signed-tag</li>
<li>tests.fetchgit.fetchTags</li>
<li>tests.fetchgit.leave-git</li>
<li>tests.fetchgit.prefetch-git-no-add-path</li>
<li>tests.fetchgit.rootDir</li>
<li>tests.fetchgit.simple</li>
<li>tests.fetchgit.simple-tag</li>
<li>tests.fetchgit.sparseCheckout</li>
<li>tests.fetchgit.sparseCheckoutNonConeMode</li>
<li>tests.fetchgit.submodule-deep</li>
<li>tests.fetchgit.submodule-leave-git</li>
<li>tests.fetchgit.submodule-leave-git-deep</li>
<li>tests.fetchgit.submodule-revision-count</li>
<li>tests.fetchgit.submodule-simple</li>
<li>tests.fetchgit.withGitConfig</li>
<li>tests.fetchpatch.decode</li>
<li>tests.fetchpatch.fileWithApostrophe</li>
<li>tests.fetchpatch.fileWithSpace</li>
<li>tests.fetchpatch.full</li>
<li>tests.fetchpatch.hunks</li>
<li>tests.fetchpatch.relative</li>
<li>tests.fetchpatch.simple</li>
<li>tests.fetchpatch2.decode</li>
<li>tests.fetchpatch2.fileWithApostrophe</li>
<li>tests.fetchpatch2.fileWithSpace</li>
<li>tests.fetchpatch2.full</li>
<li>tests.fetchpatch2.hunks</li>
<li>tests.fetchpatch2.relative</li>
<li>tests.fetchpatch2.simple</li>
<li>tests.fetchurl.flag-appending-curlOpts</li>
<li>tests.fetchurl.flag-appending-curlOptsList</li>
<li>tests.fetchurl.flag-appending-netrcPhase-curlOpts</li>
<li>tests.fetchurl.flag-appending-netrcPhase-curlOptsList</li>
<li>tests.fetchurl.hashedMirrors</li>
<li>tests.fetchurl.header</li>
<li>tests.fetchurl.no-skipPostFetch</li>
<li>tests.fetchurl.showURLs-urls-mirrors</li>
<li>tests.fetchurl.urls-mirrors</li>
<li>tests.fetchurl.urls-simple</li>
<li>tests.fetchzip.hiddenDir</li>
<li>tests.fetchzip.postFetch</li>
<li>tests.fetchzip.simple</li>
<li>tests.haskell.cabalSdist.assumptionLocalHasDirectReference</li>
<li>tests.haskell.cabalSdist.localHasNoDirectReference</li>
<li>tests.lib-tests</li>
<li>tests.pnpm.pnpm-fixup-state-db</li>
<li>tests.pnpm.pnpm-workspaces.legacy</li>
<li>tests.pnpm.pnpm-workspaces.structured</li>
<li>tests.pnpm.pnpm_11_v4</li>
<li>tests.prefer-remote-fetch.fetchFromGitHub</li>
<li>tests.prefer-remote-fetch.fetchgit</li>
<li>tests.prefer-remote-fetch.fetchurl</li>
<li>tests.prefer-remote-fetch.fetchzip</li>
<li>tests.testers.runCommand.dns-resolution</li>
<li>tests.testers.runCommand.nonDefault-hash</li>
</ul></details>

ShamrockLee committed:
fetchFromGitProvider: Dynamic: handle env to circumvent aach64-darwin infinite recursion

Assisted-by: claude-opus-4.8 and claude-sonnet-4.6 via VSCode plugin anthropic.claude-code

ShamrockLee committed:
fetchgit: respect user-provided derivationArgs.env

ShamrockLee commented:
Even with pure `fetchFromGitProviderDynamic`, the performance is not as bad as before. ($\\Delta$ gc.heapSize from +15% to  +6.6064%)

## Unchanged values

| metric                          |   value |
|---------------------------------|---------|
| sizes.Attr[^sizes.Attr]         |      16 |
| sizes.Bindings[^sizes.Bindings] |      24 |
| sizes.Env[^sizes.Env]           |       8 |
| sizes.Value[^sizes.Value]       |      16 |

## Updated values

| metric                                          |     mean_before |      mean_after |      mean_diff |   mean_pct_change |   p_value |    t_stat |
|-------------------------------------------------|-----------------|-----------------|----------------|-------------------|-----------|-----------|
| time.cpu[^time.cpu]                             |         24.6761 |         25.5663 |         0.8902 |            3.5815 |    0.0000 |    4.6591 |
| time.gc[^time.gc]                               |          2.1594 |          2.4134 |         0.2540 |           11.8316 |    0.0956 |    1.7091 |
| time.gcFraction[^time.gcFraction]               |          0.0808 |          0.0856 |         0.0048 |            7.6010 |    0.1793 |    1.3683 |
| gc.cycles[^gc.cycles]                           |          7.3590 |          7.6410 |         0.2821 |            4.0782 |    0.0004 |    3.8638 |
| gc.heapSize[^gc.heapSize]                       | 2250643403.4872 | 2379063296.0000 | 128419892.5128 |            6.6064 |    0.0000 |    6.2687 |
| gc.totalBytes[^gc.totalBytes]                   | 3669162210.8718 | 3889200370.4615 | 220038159.5897 |            6.5449 |    -      |   12.6486 |
| envs.bytes[^envs.bytes]                         |  468232847.7949 |  502216634.8718 |  33983787.0769 |            8.1831 |    -      |   12.5550 |
| list.bytes[^list.bytes]                         |   77741929.2308 |   79352710.7692 |   1610781.5385 |            2.2846 |    -      |   12.6239 |
| sets.bytes[^sets.bytes]                         | 1630921191.1795 | 1766162610.6667 | 135241419.4872 |            8.9321 |    -      |   12.5595 |
| symbols.bytes[^symbols.bytes]                   |    1670114.1026 |    1670895.5641 |       781.4615 |            0.0516 |    -      |  563.7482 |
| values.bytes[^values.bytes]                     |  734880525.9487 |  765969252.1026 |  31088726.1538 |            4.9054 |    -      |   12.6455 |
| envs.number[^envs.number]                       |   21223081.3077 |   22632840.0769 |   1409758.7692 |            7.6061 |    -      |   12.5566 |
| nrAvoided[^nrAvoided]                           |   29071357.9487 |   31385465.0769 |   2314107.1282 |            8.6935 |    -      |   12.5682 |
| nrExprs[^nrExprs]                               |    1610225.4615 |    1611033.7692 |       808.3077 |            0.0537 |    -      | 4674.7808 |
| nrFunctionCalls[^nrFunctionCalls]               |   18323366.1282 |   19352760.2308 |   1029394.1026 |            6.4425 |    -      |   12.5570 |
| nrLookups[^nrLookups]                           |    9921984.3077 |   10221048.4103 |    299064.1026 |            3.4352 |    -      |   12.4805 |
| nrOpUpdateValuesCopied[^nrOpUpdateValuesCopied] |   51563611.0000 |   54511557.8974 |   2947946.8974 |            6.1919 |    -      |   12.5709 |
| nrOpUpdates[^nrOpUpdates]                       |    4491474.6667 |    5104923.4103 |    613448.7436 |           14.4522 |    -      |   12.5501 |
| nrPrimOpCalls[^nrPrimOpCalls]                   |    9714372.4359 |   10102090.8462 |    387718.4103 |            4.4647 |    -      |   12.5539 |
| nrThunks[^nrThunks]                             |   28746491.7692 |   30564964.7179 |   1818472.9487 |            7.1643 |    -      |   12.6037 |
| sets.number[^sets.number]                       |    6828709.9231 |    7594593.9744 |    765884.0513 |           12.2105 |    -      |   12.5435 |
| symbols.number[^symbols.number]                 |     127895.4359 |     127938.8462 |        43.4103 |            0.0357 |    -      |  319.0486 |
| values.number[^values.number]                   |   45930032.8718 |   47873078.2564 |   1943045.3846 |            4.9054 |    -      |   12.6455 |
| envs.elements[^envs.elements]                   |   37306024.6667 |   40144239.2821 |   2838214.6154 |            8.5100 |    -      |   12.5542 |
| list.concats[^list.concats]                     |    1902511.2564 |    1904067.5128 |      1556.2564 |            0.0870 |    -      |   11.8429 |
| list.elements[^list.elements]                   |    9717741.1538 |    9919088.8462 |    201347.6923 |            2.2846 |    -      |   12.6239 |
| sets.elements[^sets.elements]                   |   91689509.5641 |   98993272.2051 |   7303762.6410 |            8.5961 |    -      |   12.5620 |

[^time.cpu]: Number of seconds of CPU time accounted by the OS to the Nix evaluator process. On UNIX systems, this comes from [`getrusage(RUSAGE_SELF)`](https://man7.org/linux/man-pages/man2/getrusage.2.html).
[^time.gc]: Number of seconds of CPU time accounted by the Boehm garbage collector to performing GC.
[^time.gcFraction]: What fraction of the total CPU time is accounted towards performing GC.
[^gc.cycles]: Number of times garbage collection has been performed.
[^gc.heapSize]: Size in bytes of the garbage collector heap.
[^gc.totalBytes]: Size in bytes of all allocations in the garbage collector.
[^envs.bytes]: Size in bytes of all `Env` objects allocated by the Nix evaluator. These are almost exclusively created by [`nix-env`](https://nix.dev/manual/nix/stable/command-ref/nix-env.html).
[^list.bytes]: Size in bytes of all [lists](https://nix.dev/manual/nix/stable/language/syntax.html#list-literal) allocated by the Nix evaluator.
[^sets.bytes]: Size in bytes of all [attrsets](https://nix.dev/manual/nix/stable/language/syntax.html#list-literal) allocated by the Nix evaluator.
[^symbols.bytes]: Size in bytes of all items in the Nix evaluator symbol table.
[^values.bytes]: Size in bytes of all values allocated by the Nix evaluator.
[^envs.number]: The count of all `Env` objects allocated.
[^nrAvoided]: The number of thunks avoided being created.
[^nrExprs]: The number of expression objects ever created.
[^nrFunctionCalls]: The number of function calls ever made.
[^nrLookups]: The number of lookups into an attrset ever made.
[^nrOpUpdateValuesCopied]: The number of attrset values copied in the process of merging attrsets.
[^nrOpUpdates]: The number of attrsets merge operations (`//`) performed.
[^nrPrimOpCalls]: The number of function calls to primops (Nix builtins) ever made.
[^nrThunks]: The number of [thunks](https://nix.dev/manual/nix/latest/language/evaluation.html#laziness) ever made. A thunk is a delayed computation, represented by an expression reference and a closure.
[^sets.number]: The number of attrsets ever made.
[^symbols.number]: The number of symbols ever added to the symbol table.
[^values.number]: The number of values ever made.
[^envs.elements]: The number of values contained within an `Env` object.
[^list.concats]: The number of list concatenation operations (`++`) performed.
[^list.elements]: The number of values contained within a list.
[^sets.elements]: The number of values contained within an attrset.
[^sizes.Attr]: Size in bytes of the `Attr` type.
[^sizes.Bindings]: Size in bytes of the `Bindings` type.
[^sizes.Env]: Size in bytes of the `Env` type.
[^sizes.Value]: Size in bytes of the `Value` type.

ShamrockLee commented:
IDK why the performance comparison stats wobble so much from run to run.

llakala commented:
> While gc.heapSize still has a 1.7% increase, the CPU time change is now negligible.

As you noticed later, cpuTime is extremely untrustworthy. I have some WIP work to replace it on the nixpkgs end with a `perf` counter. The non-timing stats are the deterministic ones that should be optimized for.

As it stand, this is still a massive stats regression for what seems to be little gain, unless I'm missing something.

ShamrockLee committed:
fetchFromGitProvider: shrink Simple-path retained eval heap

The Simple (non-overriding) path is what a tree-wide evaluation forces for
every `src`, so its retained live set dominates the fetcher's contribution to
peak `gc.heapSize`. Two changes bring it closer to master's plain-function
fetcher while keeping the dynamic `overrideAttrs`-switches-fetchers behaviour
(which lives in the Dynamic path and stays a lazy thunk):

- Add `extendDrvArgsSharedFor lean`. The Simple path now uses a lean variant
  that computes url/netrcPhase/name/meta eagerly from the closure and drops the
  recomputable output attributes (archiveUrl, browsableUrl, gitRepoUrl [-> passthru],
  netrcMachineName, varPrefix, pos [-> meta.position]) from `derivationArgs`.
  Since these are FODs, the leaner env changes `drvPath` but not the output hash
  or `name`, so `outPath` is unchanged and nothing rebuilds. `domain`/`private`
  are kept because fetchFromGitHub reads them off `finalAttrs` to build archiveUrl.
  The Dynamic path keeps the full (finalAttrs-based) version so overriding e.g.
  `owner` still re-derives the url.

- Collapse the no-op `fetchFromGitProviderSimple` `extendMkDerivation` wrapper
  (its `extendDrvArgs` was `mirrorArgs (args: {})`) into a raw functor attrset,
  removing one layer's retained overrideAttrs/functionArgs overhead per src.

Measured retained maxRSS: 16.0 -> 15.05 kB/src (~34% of the gap to master's
13.2 kB/src). Verified outPath/name/url byte-identical to master across
rev/tag/submodule/enterprise paths and on ripgrep.src; overrideAttrs still
switches fetcher and updates revCustom.

Assisted-by: claude-opus-4.8 via VSCode plugin anthropic.claude-code

ShamrockLee committed:
fetchFromGitProvider: fix `private = true` leaking to the fetcher

`private` was not listed in `excludeDrvArgNamesShared`, so it passed through
the choice layer to `fetchzip`/`fetchurl` as an unexpected top-level argument,
making every `private = true` fetch fail to evaluate. It is now stripped like
the other provider-only arguments; the value still reaches the derivation via
`derivationArgs.private`, so `finalAttrs.private` (used to build the archive URL
and netrc phase) is unaffected.

Non-private fetches never carried a top-level `private`, so their drvPaths are
unchanged; private fetches are FODs, so only their drvPath (not outPath) moves.

Assisted-by: claude-opus-4.8 via VSCode plugin anthropic.claude-code

ShamrockLee committed:
fetchFromGitProvider: drop tryEval guard in Simple fetcher selection

The Simple path determines `useFetchGit` by projecting the `useFetchGitArgs`
out of a `fpArgs pseudoFinalAttrs` pre-pass. The per-attribute `tryEval` guard
existed to treat any `useFetchGitArg` that references `finalAttrs` as
non-default (erring toward fetchgit). No such fixed-point `useFetchGitArg`
exists in practice — master's fetchFromGitHub has no fixed-point arguments and
`finalAttrs` is only used to let downstream attributes (urls, etc.) re-derive on
override — so the guard only adds per-invocation allocation. Dropping it removes
9 `tryEval` allocations per fetch.

Tree-wide eval of 60k srcs: gc.heapSize 889M -> 872M, totalBytes -17MB.
outPath/url and overrideAttrs fetcher-switching unchanged.

Assisted-by: Claude-opus-4.8 via VSCode extension anthropic.claude-code

ShamrockLee committed:
fetchFromGitProvider: allow pre-computed useFetchGit, expose getUseFetchGit

The Simple path determines the fetcher by projecting `useFetchGitArgs` out of a
`fpArgs pseudoFinalAttrs` pre-pass, which forces the whole (discarded) provider
output spine just to read a handful of user-provided arguments.

Factor the projection into `useFetchGitFromArgs` and let the Simple path take an
optional pre-computed `useFetchGit` (`fetchFromGitProviderSimpleFor`,
`applyWithUseFetchGit`). When it is `null` (the default `__functor`, and any
direct/generic caller) the pre-pass still runs, so behaviour is unchanged. A
downstream builder that holds the raw arguments can instead call the newly
exposed `getUseFetchGit` and pass the result to `applyWithUseFetchGit`, skipping
the pre-pass entirely — wired up for fetchFromGitHub in a follow-up.

Assisted-by: Claude-opus-4.8 via VSCode extension anthropic.claude-code

ShamrockLee committed:
fetchFromGitHub: pre-compute useFetchGit to skip the provider pre-pass

fetchFromGitHub holds its raw arguments, so it can determine the fetcher up
front instead of paying the provider's `fpArgs pseudoFinalAttrs` pre-pass (which
builds and discards the whole extendDrvArgs output spine on every fetch).

Keep the `lib.extendMkDerivation` result as `base` for its metadata
(`__functionArgs`, `expectDrvArgs`, `excludeDrvArgNames`, `extendDrvArgs`) and
override only its `__functor` to pre-compute `useFetchGit` via
`provider.getUseFetchGit` and hand it to `provider.applyWithUseFetchGit`. The
rare finalAttrs-function call form falls back to the pre-pass (`null`).

Tree-wide eval of 60k GitHub srcs: gc.heapSize 889M -> 839M (crosses a bucket),
totalBytes -49MB. outPath/url/name, backend selection (deepClone, leaveDotGit,
fetchLFS, forceFetchGit, sparseCheckout, fetchSubmodules), private, override,
overrideAttrs fetcher-switching and __functionArgs all verified unchanged.

Assisted-by: Claude-opus-4.8 via VSCode extension anthropic.claude-code

ShamrockLee committed:
fetchFromGitProvider: default to GitHub; make fetchFromGitHub an alias

Fold the GitHub-specific defaults into fetchFromGitProvider so it fetches from
GitHub out of the box, and reduce fetchFromGitHub to a plain alias. This removes
fetchFromGitHub's `lib.extendMkDerivation` layer (the `ffgh` extendDrvArgs +
merge) from every GitHub fetch; folding the urls in as lazy destructure-defaults
also avoids allocating browsableUrl/gitRepoUrl twice.

- `withUnzip ? false` module argument (default suits GitHub's tar.gz archives);
  the fetchzip `withUnzip` override moves here from fetchgithub, flip it with
  `.override { withUnzip = true; }` for providers serving `.zip`.
- GitHub defaults on `extendDrvArgsSharedFor`: providerName/githubBase/domain,
  the archive/private-API/enterprise/git urls, netrcMachineName, github-vs-generic
  purl. Defaults read `finalAttrs` so overriding re-derives. `extension` is now
  emitted top-level (a fetchzip arg; in derivationArgs it loses to fetchzip's
  `extension ? null`).
- Provider `__functor` now dispatches on the argument form: a plain attrset (the
  fetchFromGitHub case) pre-computes `useFetchGit` and skips the pre-pass; a
  finalAttrs function (a downstream extendMkDerivation layer) falls back to it.
  Non-GitHub providers keep their own `extendDrvArgs` layer.

Tree-wide eval of 60k GitHub srcs: gc.heapSize 838.9M -> 788.5M (crosses a
bucket), totalBytes -86MB. outPath/drvPath/url/name and lib.functionArgs are
byte-identical across the full matrix (rev/tag/submodule/deepClone/sparse/
enterprise/private/private-enterprise/private-fetchgit-netrc/override-switch/
override-rev) and real packages; withUnzip and .override verified.

Assisted-by: Claude-opus-4.8 via VSCode extension anthropic.claude-code

llakala commented:
Well, sending the optimizations off to the robot wasn't what I had in mind.

Anyways, the amount of ridiculous comments now for basic optimizations is embarassing. I would hope you'd have looked at the output of the commits before pushing them...

ShamrockLee commented:
> Anyways, the amount of ridiculous comments now for basic optimizations is embarassing. I would hope you'd have looked at the output of the commits before pushing them...

I acknowledge that. The diff (both code and comments) is ugly and need some good clean-up. That's one of the reason it is still a draft.

Sorry for making the noise.

ShamrockLee commented:
> Anyways, the amount of ridiculous comments now for basic optimizations is embarassing.

I got time to take a good look at how the agent optimise `fetchFromGitProviderSimple`, and it is surprisingly trivial.
In short,
1. Pass fewer arguments to the result derivation.
2. Handle plain set / fixed-point arguments by hand in `fetchFromGitProviderSimple`.

I'll adjust my initial implementation and integrate these concept.
The code should be much cleaner then.

@llakala My apology for requesting you before even reviewing the diff myself.

I only meant to ask if the overhead (+0.6% heap size and +0.7% total bytes) was acceptable.
